content/blog/2020-07-03-optimizing-dvc-remotes.md

```diff
@@ -0,0 +1,390 @@
+about the DVC cache and remote storage structure.
+
+```
+.dvc/cache
+├── 00
+│   ├── 411460f7c92d2124a67ea0f4cb5f85
+│   ├── 6f52e9102a8d3be2fe5614f42ba989
+│   └── ...
+├── 01
+├── 02
+├── 03
+├── ...
+└── ff
+```
+
+_Example DVC cache_
+
+- Files versioned by DVC are identified and stored according to their
+  [MD5](https://en.wikipedia.org/wiki/MD5) hash. The cache is organized into 256
+  subdirectories (from `00` to `ff`), where each subdirectory is the first 2
+  characters in an MD5 hash string.
+- MD5 is an
+  [evenly distributed](https://michiel.buddingh.eu/distribution-of-hash-values)
+  hash function, so the DVC cache will be evenly distributed (i.e. given a large
+  enough dataset, each cache subdirectory will contain an approximately equal
+  number of files)
+- DVC remote storage structure mirrors the DVC local cache structure, so a DVC
+  remote will also be evenly distributed
+
+## How DVC 1.0 automatic selects a query method
+
+As we established earlier, choosing the optimal query method depends on:
+
+> - The number of files that we need to query.
+> - The total number of files in the remote.
+
+In DVC, the number of files we need to query is just the number of files for a
+given project revision. So, as long as we can estimate the number of files in a
+DVC remote, we can programmatically choose the optimal query method for a remote
+operation.
+
+In DVC 1.0, we accomplish this by taking advantage of the DVC remote structure.
+The over/under remote size threshold only depends on the number of files being
+queried (i.e. the number of files in our DVC versioned dataset). And as we have
+already established, a DVC remote will be evenly distributed. Therefore, if we
+know the number of files contained in a subset of the remote, we can then
+estimate the number of files contained in the entire remote.
+
+For example, if we know that the remote subdirectory `00/` contains 10 files, we
+can estimate that the remote contains roughly `256 * 10 = 2,560` files in total.
+So, by requesting a list of one subdirectory at a time (rather than the full
+remote) via the cloud storage API, we can calculate a running estimate of the
+total remote size. If the running estimated total size goes over the threshold
+value, DVC will stop fetching the contains of the remote subdirectory, and
+switch to querying each file in our dataset individually. If DVC reaches the end
+of the subdirectory without the estimated size going over the threshold, it will
+continue to fetch the full listing for the rest of the remote.
+
+By estimating remote size in DVC 1.0, we can ensure that we always use the
+optimal method when querying remote status.
+
+_Note: The DVC remote structure also allows us to parallelize the second query
+method described earlier. By fetching the contents of each subdirectory
+individually, we can run up to 256 sets of sequential queries at once._
+
+## How DVC 1.0 reduces the number of files to query
```

Comment:
can we put here and in other section some visuals - even number of requests, etc - it's great to have more plots.